### PR TITLE
Ufs url partly pull request.

### PR DESCRIPTION
**Pull request by YichuanSun**
Reviewing on GitHub website may be more convenient.

### CODE REVIEW

File: dora/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheFileInStreamTest.java [hunk 1]
```diff
@@ -915,6 +929,17 @@ public FileInStream openFile(URIStatus status, OpenFilePOptions options)
       }
     }
 
+    @Override
+    public FileInStream openFile(UfsUrl ufsPath, OpenFilePOptions options)
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheFileInStreamTest.java:978]: why this change?

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java [hunk 2]
```diff
@@ -1,18 +1,87 @@
...
 
   public UfsUrl(UfsUrlMessage proto) {
+    Preconditions.checkArgument(proto.getPathComponentsList().size() != 0,
+        "the proto.path is empty, please check the proto first");
     mProto = proto;
   }
 
+  public UfsUrl(String ufsPath) {
+    // TODO(Tony Sun): Considering the case below:
+    //  when scheme does not exist, how to determine the scheme, or a empty scheme.
```
Comment (jiacheliu3) [inline dora/core/common/src/main/java/alluxio/uri/UfsUrl.java]: I think we have two options:
1. Assume the input path is a relative path to the Alluxio root
2. Assume the input path is local file and prefix that with "file://"

I prefer option 1

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java [hunk 3]
```diff
@@ -1,18 +1,87 @@
...
+    Preconditions.checkArgument(!ufsPath.isEmpty(),
+            "ufsPath is empty, please input a non-empty ufsPath.");
+    List<String> preprocessingPathList = Arrays.asList(ufsPath.split(SCHEME_SEPERATOR));
+
+    // TODO(Tony Sun): Does every input path contain authority?
```
Comment (jiacheliu3) [inline dora/core/common/src/main/java/alluxio/uri/UfsUrl.java]: no, many/most input paths don't have authority. You will need to get the authority from the Alluxio root UFS config.

File: dora/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheFileInStreamTest.java (hunk 1 above)
Comment (YichuanSun) [reply dora/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheFileInStreamTest.java:978]: There are many openFile function in the LocalCacheFileInStreamTest.java, I just added a ufsurl version.

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java [hunk 4]
```diff
@@ -1,18 +1,87 @@
...
+    // Empty path is excluded here.
+    Preconditions.checkArgument(indexOfFirstSlashAfterAuthority != -1,
+        "Please input a valid path.");
+    String authorityString = authorityAndPath.substring(0, indexOfFirstSlashAfterAuthority);
+    String pathString = authorityAndPath.substring(indexOfFirstSlashAfterAuthority);
+
+    // TODO(Tony Sun): Is the input of UfsUrl the same as the input of AlluxioUri? Can I use
+    //  the string test format like AlluxioURITest.java?
+
+    // TODO(Tony Sun): Do we need to handle path like '/////path/to/dir'?
```
Comment (jiacheliu3) [inline dora/core/common/src/main/java/alluxio/uri/UfsUrl.java]: We don't need to handle illegal paths like '/////path/to/dir'

I think we can perform only basic checks in the Alluxio layer. By "basic" I mean cover all corner cases `AlluxioURITest` already raises and covers.

For other strange corner cases, Alluxio passes the path to UFS to access it, and UFS will throw an exception if the path is illegal. Alluxio just needs to properly propagate the exception back to the client, so the client  knows the input param is wrong.

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java [hunk 5]
```diff
@@ -76,13 +146,28 @@ public String asStringNoAuthority() {
...
 
   public boolean isPrefix(UfsUrl another, boolean allowEquals) {
-    // TODO: implement this
+    String thisString = asString();
+    String anotherString = another.asString();
+    if (anotherString.startsWith(thisString)) {
+      if (Objects.equals(anotherString, thisString)) {
+        return allowEquals;
+      }
+      else {
```
Comment (jiacheliu3) [inline dora/core/common/src/main/java/alluxio/uri/UfsUrl.java]:
```suggestion
      } else {
```

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java [hunk 6]
```diff
@@ -76,13 +146,28 @@ public String asStringNoAuthority() {
   }
 
   public boolean isPrefix(UfsUrl another, boolean allowEquals) {
-    // TODO: implement this
+    String thisString = asString();
+    String anotherString = another.asString();
```
Comment (jiacheliu3) [inline dora/core/common/src/main/java/alluxio/uri/UfsUrl.java]: this is a bit extra expensive. Try optimize this in your later PRs.

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java [hunk 7]
```diff
@@ -105,10 +191,28 @@ public UfsUrl getChildURL(String childName) {
...
+  public String getName() {
+    List<String> pathComponents = getPathComponents();
+    return pathComponents.get(pathComponents.size() - 1);
+  }
+
+  //TODO(Tony Sun): Add isAncestorOf() method.
+
+  public boolean isRoot() {
+    // TODO(Tony Sun): In AlluxioURI here remains judging hasAuthority(), why?
+    return getFullPath().equals(PATH_SEPERATOR) || getFullPath().isEmpty();
```
Comment (jiacheliu3) [inline dora/core/common/src/main/java/alluxio/uri/UfsUrl.java]: This is a good question. If we keep the full path in UFS here, it's not so obvious whether a path is equal to Alluxio root UFS path (or even a parent of Alluxio root UFS path). 

I think we need special optimizations for that. For now just write your own 1st version, it's fine.

That may be related to https://github.com/Alluxio/alluxio/pull/17839 and will require future code changes

File: dora/shell/src/main/java/alluxio/cli/fs/command/LsCommand.java [hunk 8]
```diff
@@ -285,13 +287,46 @@ private void ls(AlluxioURI path, boolean recursive, boolean forceLoadMetadata, b
       return;
     }
 
+//    List<URIStatus> statusListFromUfsUrl = mFileSystem.listStatus();
```
Comment (jiacheliu3) [inline dora/shell/src/main/java/alluxio/cli/fs/command/LsCommand.java:290]: remove commented code
